File: sakila-store-read-service/src/main/java/com/example/sakila/config/JPAConfig.java
package com.example.sakila.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.dao.annotation.PersistenceExceptionTranslationPostProcessor;
import org.springframework.orm.jpa.JpaTransactionManager;
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean;
import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import javax.persistence.EntityManagerFactory;
import javax.sql.DataSource;
import java.util.Properties;

@Configuration
@EnableTransactionManagement
public class JPAConfig {

  @Bean
  public LocalContainerEntityManagerFactoryBean localContainerEntityManagerFactoryBean(DataSource dataSource) {
    LocalContainerEntityManagerFactoryBean entityManagerFactoryBean = new LocalContainerEntityManagerFactoryBean();
    entityManagerFactoryBean.setDataSource(dataSource);
    entityManagerFactoryBean.setPackagesToScan("com.example.sakila");
    entityManagerFactoryBean.setJpaVendorAdapter(new HibernateJpaVendorAdapter());
    entityManagerFactoryBean.setJpaProperties(hibernatePsql10Properties());
    return entityManagerFactoryBean;
  }

  @Bean
  public PlatformTransactionManager transactionManager(EntityManagerFactory entityManagerFactory) {
    JpaTransactionManager transactionManager = new JpaTransactionManager();
    transactionManager.setEntityManagerFactory(entityManagerFactory);
    return transactionManager;
  }

  @Bean
  public PersistenceExceptionTranslationPostProcessor persistenceExceptionTranslator() {
    return new PersistenceExceptionTranslationPostProcessor();
  }

  private Properties hibernatePsql10Properties() {
    Properties properties = new Properties();
    properties.setProperty("hibernate.hbm2ddl.auto", "update");
    properties.setProperty("hibernate.dialect", "org.hibernate.dialect.PostgreSQL10Dialect");
    return properties;
  }
}

File: sakila-payment-service/src/main/java/com/example/sakila/config/JPAConfig.kt
package com.example.sakila.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.dao.annotation.PersistenceExceptionTranslationPostProcessor
import org.springframework.orm.jpa.JpaTransactionManager
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean
import org.springframework.orm.jpa.vendor.HibernateJpaVendorAdapter
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.annotation.EnableTransactionManagement
import java.util.*
import javax.persistence.EntityManagerFactory
import javax.sql.DataSource

@Configuration
@EnableTransactionManagement
class JPAConfig {

  @Bean
  fun localConatinerEntityManagerFactoryBean(dataSource: DataSource): LocalContainerEntityManagerFactoryBean {
    val factoryBean = LocalContainerEntityManagerFactoryBean()

    factoryBean.dataSource = dataSource
    factoryBean.setPackagesToScan("com.example.sakila")
    factoryBean.jpaVendorAdapter = HibernateJpaVendorAdapter()
    factoryBean.setJpaProperties(hibernatePsql10Properties())

    return factoryBean
  }

  @Bean
  fun persistenceExceptionTranslationPostProcessor(): PersistenceExceptionTranslationPostProcessor =
      PersistenceExceptionTranslationPostProcessor()

  @Bean
  fun transactionManager(entityManagerFactory: EntityManagerFactory): PlatformTransactionManager {
    val transactionManager = JpaTransactionManager()

    transactionManager.entityManagerFactory = entityManagerFactory

    return transactionManager
  }

  private fun hibernatePsql10Properties(): Properties {
    val properties: Properties = Properties()
    properties.setProperty("hibernate.dialect", "org.hibernate.dialect.PostgreSQL10Dialect")
    return properties
  }
}
